### PR TITLE
DEV: changelog: make title processing more robust

### PR DESCRIPTION
**Pull request by lucascolley**
Hey @charris,

I made some adjustments to this script (which was copied into SciPy) over there a little while ago, to make the processing of titles more robust. In particular:

- a bug is fixed with the addition of ellipses for long remainders
- automatic handling of single backticks to double backticks is added (so you don't need to worry about editing titles which use single backticks :) )
- asterisks are also sanitised (I think that change pre-dated my work)
- code comments are added to make this more maintainable going forward

I haven't tested this, so it might be broken currently - would you be able to give it a test?

### CODE REVIEW

File: tools/changelog.py [hunk 1]
```diff
@@ -132,15 +132,47 @@ def main(token, revision_range):
...
+                # use the first 60 characters and the next word
                 title = title[:60] + remainder
-        print(pull_msg.format(pull.number, pull.html_url, title))
+
+            if title.count('`') % 4 != 0:
+                # ellipses have cut in the middle of a code block,
+                # so finish the code block before the ellipses
+                title = title[:-3] + '``...'
+
+        msg = msg.format(pull.number, pull.html_url, title)
```
[the remark this answers is not in the document]
Comment (lucascolley) [reply tools/changelog.py]: oops, fixed

**Comment by charris** [comment]
The single backtick replacement isn't working. It can be a bit tricky, especially at line ends. I use vim search and replace when there a bunch, and it is tricky there also, I use before and after groups.

**Comment by lucascolley** [comment]
> The single backtick replacement isn't working

could you give a reproducer for which this fails?

**Comment by charris** [comment]
Seems quite general:
```
* `#27728 <https://github.com/numpy/numpy/pull/27728>`__: BUG: Handle `--lower` for F2PY directives and callbacks
* `#27729 <https://github.com/numpy/numpy/pull/27729>`__: BUG: f2py: fix issues with thread-local storage define
* `#27730 <https://github.com/numpy/numpy/pull/27730>`__: TST: Add an F2PY check for exposing variables without functions
* `#27731 <https://github.com/numpy/numpy/pull/27731>`__: BUG: Fix `fortranname` for functions
* `#27734 <https://github.com/numpy/numpy/pull/27734>`__: Fix documentation for the chi-square distribution
* `#27735 <https://github.com/numpy/numpy/pull/27735>`__: ENH: Add a `__dict__` to ufunc objects and allow overriding `__doc__`
```

**Comment by charris** [comment]
Hmm, looks like I don't have the pr correctly checked out...

**Comment by lucascolley** [comment]
(fyi I can't see whether the original titles had single or double backticks without triage rights, so input and output would be useful to debug!)

**Comment by charris** [comment]
OK, to start with I see
```
/home/charris/Workspace/numpy.git/tools/changelog.py:138: SyntaxWarning: invalid escape sequence '\ '
  post = '\ ' + matchobj.group(2)
/home/charris/Workspace/numpy.git/tools/changelog.py:151: SyntaxWarning: invalid escape sequence '\g'
  "\g<pre>``\g<post>",
```

Otherwise seems to work. Maybe want raw strings?

**Comment by lucascolley** [comment]
made them raw strings - let me know if that still works!

**Comment by charris** [comment]
Looks good. Thanks @lucascolley .

**Comment by charris** [comment]
Found two nits, which I will correct with the next changelog generation.

**Comment by lucascolley** [comment]
> Found two nits, which I will correct with the next changelog generation.

Please ping me if you remember, would be good to fix in SciPy too!

**Comment by charris** [comment]
Maybe just one
```
`#26944 <https://github.com/numpy/numpy/pull/26944>`__: TYP: Annotate ``numpy._core._type_aliases``\ .
```
Perhaps `r'\ '` should just be `' '`.

**Comment by lucascolley** [comment]
Did it work before I made it a raw string? Was it a linter giving you the `SyntaxWarning`s?

**Comment by charris** [comment]
> Did it work before I made it a raw string? Was it a linter giving you the SyntaxWarnings?

Don't know. No linter, just ran `python tools/changelog.py ...`.

**Comment by charris** [comment]
However, removing the backslash fixes the problem.,
```
-* TYP: Annotate ``numpy._core._type_aliases``\ .
+* TYP: Annotate ``numpy._core._type_aliases`` .
```

The other possibility glitch was `\*`, but it renders without the backslash when not in code.

**Comment by charris** [comment]
\ . \*.

**Comment by lucascolley** [comment]
> However, removing the backslash fixes the problem.,
> 
> ```
> -* TYP: Annotate ``numpy._core._type_aliases``\ .
> +* TYP: Annotate ``numpy._core._type_aliases`` .
> ```
> 
> The other possible glitch was \*, but it renders without the backslash when not in code.

But how does rST render the backlash-space combination? It may be that producing a string with a backslash in is intentional.

**Comment by charris** [comment]
Rst renders the space without the backslash, but also works fine if the backslash is omitted.

**Comment by lucascolley** [comment]
> Rst renders the space without the backslash, but also works fine if the backslash is omitted.
> 
> EDIT: At least that is how GitHub renders the rst file.

If there was no space in the input, then there shouldn't be a space in the output. Am I correct in thinking that we're adding in an extra space that shouldn't be there? What happens if you go back to before we made the string raw?

**Comment by charris** [comment]
```
In [1]: r'\ '
Out[1]: '\\ '

In [2]: '\ '
<>:1: SyntaxWarning: invalid escape sequence '\ '
<ipython-input-2-f3cdf6e5e445>:1: SyntaxWarning: invalid escape sequence '\ '
  '\ '
Out[2]: '\\ '
```

**Comment by lucascolley** [comment]
So for this example:

> TYP: Annotate ``numpy._core._type_aliases`` .

Did the original title contain a space before the period?

**Comment by charris** [comment]
No.

**Comment by lucascolley** [comment]
Actually, scratch that, I think the state of this PR is fine. GitHub renders `a\ b` in rST as `ab`.

**Pull request merged**